File: library/src/main/java/com/github/library/BaseClient.java
package com.github.library;

import android.annotation.SuppressLint;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

import androidx.collection.ArrayMap;

import com.github.library.enums.AuthType;
import com.github.library.helper.LogHelper;
import com.github.library.model.AuthModel;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.lang.ref.WeakReference;
import java.util.concurrent.TimeUnit;

import okhttp3.Call;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

abstract class BaseClient {

    WeakReference<Context> appContext;

    String clientId;
    String clientSecret;
    String site;

    String token;

    String grantType;

    String username;
    String password;

    AuthType authType;

    ArrayMap<String,String> headers;

    int timeMilliSecond = 60;

    boolean debugEnable             = true;

    private static OkHttpClient instance;

    static synchronized OkHttpClient getClient(int timeOut, boolean enableDebug) {
        if (instance == null) {
            instance = new OkHttpClient()
                    .newBuilder()
                    .connectTimeout(timeOut, TimeUnit.MILLISECONDS)
                    .build();

            if (enableDebug) {
                instance.newBuilder()
                        .addInterceptor(new LoggingInterceptor())
                        .build();
            }
        }
        return instance;
    }

    public void cancelAllRequest() {
        for (Call call : getClient(timeMilliSecond,debugEnable).dispatcher().queuedCalls()) {
            call.cancel();
        }
    }

    public void cancelCallWithTag(String tag) {
        for (Call call : getClient(timeMilliSecond,debugEnable).dispatcher().queuedCalls()) {
            try {
                Object item = call.request().tag();
                if (item != null){
                    if (item.equals(tag))
                        call.cancel();
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    private static class LoggingInterceptor implements Interceptor {

        LogHelper logHelper = new LogHelper(RestClient.class);

        @SuppressLint("DefaultLocale")
        @NotNull
        @Override
        public Response intercept(@NotNull Interceptor.Chain chain) throws IOException {
            Request request = chain.request();
            long t1 = System.nanoTime();
            logHelper.d(String.format("Sending request %s on %s%n%s", request.url(), chain.connection(), request.headers()));
            Response response = chain.proceed(request);
            long t2 = System.nanoTime();
            logHelper.d(String.format("Received response for %s in %.1fms%n%s", response.request().url(), (t2 - t1) / 1e6d, response.headers()));
            return response;
        }
    }

    boolean checkNetworkConnection(Context context) {
        try {
            if (context == null) {
                return false;
            }
            ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
            if (connectivityManager == null) {
                return true;
            }
            NetworkInfo nwInfo = connectivityManager.getActiveNetworkInfo();
            return nwInfo == null || !nwInfo.isConnectedOrConnecting();
        } catch (Exception ex) {
            return true;
        }
    }

    AuthModel getAuthModel(){
        AuthModel auth = new AuthModel();
        auth.setClientId(this.clientId);
        auth.setClientSecret(this.clientSecret);
        auth.setSite(this.site);
        auth.setToken(this.token);
        auth.setGrantType(this.grantType);
        auth.setUsername(this.username);
        auth.setPassword(this.password);
        auth.setAuthType(this.authType);
        auth.setHeaders(this.headers);
        return auth;
    }
}

File: library/src/main/java/com/github/library/BaseClient.kt
package com.github.library

import android.content.Context
import android.net.ConnectivityManager
import androidx.collection.ArrayMap
import com.github.library.enums.AuthType
import com.github.library.helper.LogHelper
import com.github.library.model.AuthModel
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response
import java.io.IOException
import java.lang.Exception
import java.lang.ref.WeakReference
import java.util.concurrent.TimeUnit


abstract class BaseClient {

    val log = LogHelper(this::class.java)

    var appContext: WeakReference<Context>? = null

    var clientId: String = ""
    var clientSecret: String = ""
    var site: String = ""

    var token: String = ""

    var grantType: String = ""

    var username: String = ""
    var password: String = ""

    var baseUrl: String = ""

    var authType: AuthType = AuthType.NO_AUTH

    var headers: ArrayMap<String, String>? = null

    var timeMilliSecond = 60

    var debugEnable = true

    val authModel: AuthModel get() {
            val auth = AuthModel()
            auth.clientId = this.clientId
            auth.clientSecret = this.clientSecret
            auth.site = this.site
            auth.token = this.token
            auth.grantType = this.grantType
            auth.username = this.username
            auth.password = this.password
            auth.authType = this.authType
            auth.headers = this.headers
            return auth
        }

    private class LoggingInterceptor : Interceptor {

        internal var logHelper = LogHelper(RestClient::class.java)

        @Throws(IOException::class)
        override fun intercept(chain: Interceptor.Chain): Response {
            val request = chain.request()
            val t1 = System.nanoTime()
            logHelper.d(String.format("Sending request ${request.url()} on ${chain.connection()}%n${request.headers()}"))
            val response = chain.proceed(request)
            val t2 = System.nanoTime()
            logHelper.d(String.format("Received response for ${response.request().url()} in ${(t2 - t1) / 1e6} %n ${response.headers()}"))
            return response
        }
    }

    fun checkNetworkConnection(context: Context?): Boolean {
        context ?: return false
        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val nwInfo = connectivityManager.activeNetworkInfo ?: return false
        return nwInfo.isConnected
    }

    companion object {

        private var instance: OkHttpClient? = null

        fun getClient(timeOut: Int, enableDebug: Boolean): OkHttpClient? {
            if (instance == null) {
                instance = OkHttpClient()
                        .newBuilder()
                        .connectTimeout(timeOut.toLong(), TimeUnit.SECONDS)

                        .build()

                if (enableDebug) {
                    instance!!.newBuilder()
                            .addInterceptor(LoggingInterceptor())
                            .build()
                }
            }
            return instance
        }
    }

    fun cancelAllRequest() {
        for (call in getClient(timeMilliSecond, debugEnable)?.dispatcher()?.runningCalls()!!) {
            call.cancel()
        }
    }

    fun cancelCallWithTag(tag: String) {
        for (call in getClient(timeMilliSecond, debugEnable)?.dispatcher()?.runningCalls()!!) {
            try {
                if (call.request().tag()?.equals(tag)!!) {
                    call.cancel()
                }
            } catch (e: Exception) {
                log.e(e = e)
            }
        }
    }

}
